src/Week10/AParser.hs

```diff
@@ -0,0 +1,172 @@
+satisfy :: (Char -> Bool) -> Parser Char
+satisfy p = Parser f
+  where
+    f [] = Nothing    -- fail on the empty input
+    f (x:xs)          -- check if x satisfies the predicate
+                        -- if so, return x along with the remainder
+                        -- of the input (that is, xs)
+        | p x       = Just (x, xs)
+        | otherwise = Nothing  -- otherwise, fail
+
+-- Using satisfy, we can define the parser 'char c' which expects to
+-- see exactly the character c, and fails otherwise.
+char :: Char -> Parser Char
+char c = satisfy (== c)
+
+{- For example:
+
+*Parser> runParser (satisfy isUpper) "ABC"
+Just ('A',"BC")
+*Parser> runParser (satisfy isUpper) "abc"
+Nothing
+*Parser> runParser (char 'x') "xyz"
+Just ('x',"yz")
+
+-}
+
+-- For convenience, we've also provided a parser for positive
+-- integers.
+posInt :: Parser Integer
+posInt = Parser f
+  where
+    f xs
+      | null ns   = Nothing
+      | otherwise = Just (read ns, rest)
+      where (ns, rest) = span isDigit xs
+
+------------------------------------------------------------
+-- Your code goes below here
+------------------------------------------------------------
+
+{-
+Exercise 1
+First, you’ll need to implement a Functor instance for Parser.
+Hint: You may find it useful to implement a function first :: (a -> b) -> (a,c) -> (b,c)
+-}
+
+first :: (a -> b) -> (a, c) -> (b, c)
+first f (a, c) = (f a, c)
+
+instance Functor Parser where
+    fmap f p = Parser $ fmap (first f) . runParser p
+    
+{-
+Exercise 2
+Now implement an Applicative instance for Parser:
+• pure a represents the parser which consumes no input and suc- cessfully returns a result of a.
+• p1 <*> p2 represents the parser which first runs p1 (which will consume some input and produce a function), 
+then passes the remaining input to p2 (which consumes more input and produces some value), 
+then returns the result of applying the function to the value. 
+However, if either p1 or p2 fails then the whole thing should also fail (put another way, p1 <*> p2 only succeeds if both p1 and p2 succeed).
+So what is this good for? Recalling the Employee example from class,
+type Name = String
+ data Employee = Emp { name :: Name, phone :: String }
+we could now use the Applicative instance for Parser to make an employee parser from name and phone parsers. 
+That is, if
+ parseName  :: Parser Name
+ parsePhone :: Parser String
+then
+ Emp <$> parseName <*> parsePhone :: Parser Employee
+is a parser which first reads a name from the input, then a phone number, and returns them combined into an Employee record. 
+Of course, this assumes that the name and phone number are right next to each other in the input, with no intervening separators. 
+We’ll see later how to make parsers that can throw away extra stuff that doesn’t directly correspond to information you want to parse.
+-}
+
+instance Applicative Parser where
+    pure a = Parser $ (\s -> Just (a, s))
+    p1 <*> p2 = Parser p
```

Comment:
This is just educational.

Something Jan suggested me long time ago. I think it makes this really easy to understand the intuition behind this. If you rewrite this without any combinators, you'll find that there's not much going on behind the scenes.
```
instance Applicative Parser where
  pure a = Parser $ \x -> Just (a, x)
  Parser f1 <*> Parser f2 = Parser $ \input -> case f1 input of
      Nothing -> Nothing
      Just (f, intermediateOutput) -> case f2 intermediateOutput of
          Nothing               -> Nothing
          Just (a, finalOutput) -> Just (f a, finalOutput)
```

Reply:
Another method for defining multiplicative which a lot of folks find simpler is by defining liftA2/map2 like function first.

```haskell
liftA2 :: (a -> b -> c) -> Parser a -> Parser b -> Parser c
```

then `(<*>) = liftA2 ($)`.

That way one doesn't need to think about currying too much for definition itself and definition of <*> is mechanical. Also to some it makes it clearer that application is in fact about combining.

To confirm this from ghci just do:

```
λ import Control.Applicative
λ x = lifetA2 ($) -- x is our custom definition of <*>
λ :t x -- check the type of our x
x :: Applicative f => f (a -> c) -> f a -> f c
λ :t (<*>) -- check the <*> to see it's the same (just shows b instead of c)
(<*>) :: Applicative f => f (a -> b) -> f a -> f b
```